external/swloop/swloop.cpp

```diff
@@ -0,0 +1,533 @@
+#include <complex>
+#include "mex.h"
+#include "Eigen/Core"
+#include "Eigen/Cholesky"
+#include "Eigen/Eigenvalues"
+#include <iostream>
+#include <vector>
+#include <thread>
+#include <atomic>
+
+typedef Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> RowMatrixXd;
+typedef std::complex<double> cd_t;
+typedef std::tuple<std::complex<double>, Eigen::VectorXcd> ev_tuple_t;
+
+struct pars {
+    bool hermit;
+    bool formfact;
+    bool incomm;
+    bool helical;
+    bool bq;
+    bool field;
+    double omega_tol;
+    size_t nMagExt;
+    size_t nTwin;
+    size_t nHkl;
+    size_t nBond;
+    size_t nBqBond;
+};
+
+struct swinputs {
+    const double *hklExt;
+    const std::complex<double> *ABCD;
```

Comment:
perhaps some comments here to explain the input or better variable names (maybe you want to keep variable names for consistency elsewhere - though there's anrgument to change them everywhere!) but not clear to be what ABCD represents!)

Reply:
`ABCD` is the flattened non-q-dependent part of the Hamiltonian; I used the same variable name as in [spinwave.m](https://github.com/SpinW/spinw/blob/master/swfiles/%40spinw/spinwave.m#L736). In terms of the theory it's the flattened matrix in eq (25) of the [arxiv paper](https://arxiv.org/pdf/1402.6069.pdf) with Fourier transform of the exchange terms factories out; in this case the Hamiltonian matrix can be expressed as:

```
[ A B;
  C D ]
```

where `A` is `A-C` in equation (26), `B` is `B`, `C` is `adjoint(B)`, and `D` is `adjoint(A)-C`.